### ISSUE TITLE
Preparation for next ray workflow run on Delta

### ISSUE THREAD

**Issue by julietcohen** [PermafrostDiscoveryGateway/viz-workflow#17]
* PDG Workflow version: ray
* Python version: 3.9
* Server: Delta

## To do before next ray workflow run on Delta:

- [X] **~~transfer IWP files (those that are the output of the pre-processing work and the input to the post-processing ray workflow) from the old allocation `bbki` in Kastan's scratch dir to projects directory on old allocation because the projects directory is not wiped after 30 days of inactivity like scratch is~~**
- **see [comment](https://github.com/PermafrostDiscoveryGateway/viz-workflow/issues/17#issuecomment-1382243864) below**
- only need to keep non-empty IWP files, since many of them are blank we can omit those from the transfer, but there may be more we do not need to keep either (neither team is sure which other IWP files are not needed)
- not necessary to transfer anything from Robyn's scratch directory in `bbki` to projects
- not necessary to keep anything that is the _output of the post-rocessing ray workflow_ in Kastan's directory either, because we are re-doing the entire thing
- plan to do this transfer with `rsync`
- [x] **make help ticket for Delta team to ask:**
1. how long the `bbki` allocation will persist since it expired
2. why `bbki` expired (number of hours does not seem likely based on output of 'accounts' and 'quota' commands, could be amount of storage or number of files?), Kastan might have insight about this
- [x] **request extra storage and higher limit for number of files in new allocation (`bbou`) like we did for `bbki`?**
- extra storage does not need to be in place for the next run in February, but the request should be submitted ASAP to secure extra storage beyond February 
- if we do _not_ get extra storage and higher limit for number of files beyond February, consider transferring output files from workflow off Delta completely after each workflow run
- [x] **determine if we should we use GPU or CPU**
- [x] Add `Date` attribute to footprints
- [x] Determine why Delta computation time is unpredictable / fix lack of parallelization in rasterization
- [x] Determine how to execute staging to rpoduce the expected amount of staged files based on the amount of input shp files (see [issue#19](https://github.com/PermafrostDiscoveryGateway/viz-workflow/issues/19)).
- [x] Determine if clipping to footprint is executing correctly (output of test run show bright/overlapping polygon strips at edges of tiles (see [viz-workflow issue#19](https://github.com/PermafrostDiscoveryGateway/viz-workflow/issues/19#issuecomment-1458905397) and [viz-staging issue#19](https://github.com/PermafrostDiscoveryGateway/viz-staging/issues/19))
- [x] Convert approved ACCESS credits to more GPU and storage

## Notes about next run on Delta:

- high ice only, like previous runs
- Elias will provide new IWP files with more thorough cleaning (masking for coastal water, lakes, glaciers, and infrastructure)
- Elias will provide the same IWP footprints as before, but with naming and directory structure that matches the IWP files
- might meet with Elias/Chandi/Robyn/Juliet to clarify matching IWP files to footprints

**Comment by julietcohen** [comment]
Ran out of disk space in `/projects/bbki/julietcohen/` when tried to `rsync` all IWP input files in a tmux session from `/scratch/bbki/kastanday/maple_data_xsede_bridges2/glacier_water_cleaned_shp`.

Command used from dir `/scratch/bbki/kastanday/maple_data_xsede_bridges2/glacier_water_cleaned_shp`:
```
(pdg_workflow) [julietcohen@dt-login02 glacier_water_cleaned_shp]$ rsync -av . /projects/bbki/julietcohen/iwp_files_2023_01_12
```

Successfully transfered 1.16 T / 4.81 T, then output this error:

```
rsync: write failed on "/projects/bbki/julietcohen/iwp_files_2023_01_12/high_ice/canada/114_115_iwp/WV03_202007232│·····
12757_104001005F708900_20JUL23212757-M1BS-504566727030_01_P002_u16rf3413_pansh/WV03_20200723212757_104001005F70890│·····
0_20JUL23212757-M1BS-504566727030_01_P002_u16rf3413_pansh.shp": Disk quota exceeded (122)                         │·····
rsync error: error in file IO (code 11) at receiver.c(374) [receiver=3.1.3] 
```

**Comment by julietcohen** [comment]
The normal limit for the `/projects` dir for an allocation is 500 GB, according to the [Delta User Guide](https://wiki.ncsa.illinois.edu/display/DSC/Delta+User+Guide):
<img width="582" alt="image" src="https://user-images.githubusercontent.com/88209596/212397235-78c29251-3752-452b-ba84-5903d7fb531b.png">

But I hoped that the Delta team had expanded this dir in `bbki` enough for our needs just as they had expanded the data storage limit for `/scratch` when Kenton requested it before Delta was in production. Unfortunately, `/projects` does not have enough space to perpetually host all our input IWP files for the ray workflow, and they cannot perpetually live in `/scratch` because eventually that will be purged after 30 days of inactivity. Hopefully, Delta will stand true to their note in the above chart that the purging is _not_ currently enabled, and we will receive a 30-day warning during which we can transfer all IWP files from `/scratch` to a location off Delta if we want to keep them. Hopefully we can just delete the entire directory after Elias transfers the new batch of cleaner IWP files.

**Comment by julietcohen** [comment]
After consulting Kastan, we likely do _not_ need to use ACCESS credits to request extra storage on `bbou` before running the IWP workflow using `bbou` credits, because we should just transfer the written files from `/tmp` to `/scratch/bbki/`, which should be possible even though `bbki` credits were not used to create the files. We likely have unlimited storage in `/scratch/bbki` so we should just use that until it expires in May.

**Comment by julietcohen** [comment]
## Updates on our data storage limits in allocations bbki and bbou:

- turns out that the unlimited scratch space we have been allowed is due to a bug in the Delta file system software stack, and all scratch directories have unlimited space at the moment
- the Delta team hopes to start enforcing the scratch quotas in about a month, at which point the quota for bbki will be the standard 1 T and the quota for bbou will be ~4.9 T

So we can retain IWP files in either bbki or bbou scratch directories for the next month without worrying about reaching storage capacity in either. However, we have to move our IWP files off of Delta completely within the next month, unless we communicate with them in advance about granting us special privileges to retain more than ~4.9 T worth of data in bbou. From their recent message, it seems like the Delta team will work with us to ensure our data is not removed without sufficient warning.

When we have the estimate for the amount of storage we will need after a complete IWP run, we can ask the Delta team if that is a reasonable amount of storage allowance after they start to enforce the scratch storage limits. If it's too much data for the Delta team to allow, I will start looking into non-Delta data storage options for after 1 month from now :grimacing: Elias should be sending over a new batch of IWP input files for the ray workflow in a week or so, which should be fewer files than before (due to removing duplicates, empty files, and some Canada regions). Hopefully having that new batch of IWP input files will help us get an accurate estimate for all outputs.

Message from the Delta team:
>Hi Juliet,
   You asked a lot in the last email and there are a few things I needs clear
up. First, due to a bug in the file system software stack we are not currently
enforcing quotas on scratch. As a result some projects, yours included, have
gotten quite a long way over their actual quota. You can use the "quota"
command on Delta to list your personal home quota and your group based quotas
on /projects and /scratch. We hope to be able to turn on quota enforcement in
about a month at which point we will have to work with teams over quota to get
usage back into a reasonable state.
Your two projects currently have the following scratch quotas and usage:
lfs quota -h -g delta_bbki /scratch
Disk quotas for grp delta_bbki (gid 17526):
>  Filesystem    used   quota   limit   grace   files   quota   limit grace
>    /scratch     83T*  1000G  1.074T       - 137931471* 1700000 1870000   
lfs quota -h -g delta_bbou /scratch
Disk quotas for grp delta_bbou (gid 17635):
>  Filesystem    used   quota   limit   grace   files   quota   limit grace
>    /scratch   34.5G  4.883T  5.371T       -  169735  8500000 9350000     
So project bbou actually has a larger, 4.9 TB quota versus the default 1TB
quota for bbki. However, usage under bbki is 83TB. That is ok for now, but
will be a problem once quota enforcement is turned back on. When that occurs
if your project remains over quota you will not be able to write new data to
scratch under that group space.
Please go ahead and keep your runs going, but do start to think about reducing
the overall scratch usage if possible.

**Comment by julietcohen** [comment]
## To address the 4th to-do item: determine if we should we use GPU or CPU

Use CPU instead of GPU when we have ample hours of CPU considering the amount of nodes planned for that run. CPU is significantly cheaper, but has about half of the amount of storage on `/tmp` compared to GPU. This is a consideration for the files that get transferred to and written to `/tmp`:
- footprints
- staged files
- geotiff files

Also adding this information to the IWP workflow documentation.

## Plan to retain more than 5.9 TB of data on Delta after our unlimited storage capacity expires

1. obtain more ACCESS credits from the same NSF grant we already have
We only have ~12,000 out of the original 200,000 credits from the grant, only 6%. Kastan used ~187,000 to buy the `bbou` allocation's 40,000 CPU hours, 2,000 GPU hours, and 4.9 TB of storage. 
2. use remaining ~12,000 and the newly obtained ACCESS credits to buy more TB of storage on `bbou` so we do not have to find an alternative data storage option besides Delta
Need to submit this request in advance of the unlimited storage options running out so it has time to process!

Kastan suggested keeping some files (the staged files, perhaps) within the 1 TB we have on `bbki` to at least utilize that space moving forward. That brings our total future storage to 5.9 TB if we do not obtain more.

**Comment by julietcohen** [comment]
# Adding `Date` attribute to footprint files

Used script on Delta `/u/julietcohen/add_date_attribute_footprints.py`

<details>
  <summary>add_date_attribute_footprints.py</summary>

```
# 1. Read in all footprint files for high, medium, and low directories
# 2. Add 'Date' attribute as YYYYMMDD pulled from filename
# 3. Save this modified footprint file to new footprint directory 

# note: the subdirectories in the new footprints directory have already been created
# because Delta does not allow files written to /scratch to also 
# create the necessary sudirectories. This was done in the command line:
# 1. `find . -type d > /u/julietcohen/dirs.txt` in original footprints dir
# 2. `xargs mkdir -p < /u/julietcohen/dirs.txt` in new footprints dir

# environment used: /scratch/bbou/julietcohen/venv/iwp_2

import geopandas as gpd
import os
import glob
from pathlib import Path

input_dir = Path('/scratch/bbou/julietcohen/IWP/footprint_files_without_date_20230119')
shp_files_posix = sorted(input_dir.rglob('*.shp'))

print('removing "PosixPath()" part of paths')

# remove prefix "PosixPath()" part from each full path
shp_files = []
for file in shp_files_posix:
    # extract just str filepath
    file_str = str(file)
    shp_files.append(file_str)

number_shp_files = len(shp_files)
new_directory = '/scratch/bbou/julietcohen/IWP/footprint_files_with_date_20230119/'

print(f'starting loop to save {number_shp_files} new footprint files to {new_directory}')

for file in shp_files:
    # save the filepath ONLY starting at high/... or medium/... or low/...
    filepath_to_keep = file[68:] 
    # read in basename that contains date
    filename = os.path.basename(file)
    # parse out date from basename
    YYYYMMDD = filename[5:13]
    print(f'reading file {file}')
    gdf = gpd.read_file(file) 
    # add date as column in gdf
    gdf['Date'] = YYYYMMDD
    # save file to new directory
    new_path = new_directory + filepath_to_keep
    gdf.to_file(new_path)
    print(f'wrote new file {new_path}')

print('Script complete.')

```
</details>

**Original footprints dir** is 51.42 MB and contains 217,552 files, 27,194 of which are `.shp` files. Simple math reveals that  1/8 files are shapefiles, and each shapefile has 7 dependency files. Looking at one subdirectory, this checks out:
<img width="592" alt="image" src="https://user-images.githubusercontent.com/88209596/217656032-9d1cbbd9-cc4d-40fc-851d-e9a085efbc0b.png">

**New footprints dir** is 27.08 MB and contains 135,970 files, 27,194 of which are `.shp` files. Simple math reveals that 1/5 files are shapefiles, and each shapefile has 4 dependency files. 

Taking a look at one subdirectory, this checks out:
<img width="596" alt="image" src="https://user-images.githubusercontent.com/88209596/217656369-a87f9c35-e9aa-4de4-a880-a20f8f235361.png">

I suppose difference in dependency files is the result of Elais' files being produced by ArcGIS (I assume), and when I added the `Date` attribute to each geodataframe and re-wrote the shapefile to a new directory, it created fewer dependency files for each shapefile. The new files lack a `.sbn` file, a `.sbx` file, and a `shp.xml` file. 

@robyngit when you have a free moment, please let me know if you think the absence of these 3 dependency files will be an issue. I do not think it will be. I would be happy to try again to add the `Date` attribute to each shapefile in a different way, without reading it and modifying it with geopandas!

**Comment by robyngit** [comment]
@julietcohen I honestly don't know much about these filetypes that are part of a shapefile. If you can read the new footprints into Geopandas and if they have a CRS, then I think it's ok. It might just have to do with the difference between how Geopandas saves shapefiles vs how ArcGIS does it, like you suggested. Just by the way, geopackage files are much nicer to deal with for these sorts of reasons. If you go through this process again, it's okay to save the footprints as .gpkg if you prefer. You just have to set the `ext_footprint` option in the config appropriately.

**Comment by julietcohen** [comment]
@robyngit Awesome, thank you! Yes, the new shapefiles have the same CRS as the old ones. I'll move forward with these.

That's a good point that I could have just made them geopackages. I can always go back and do that with a slight adjustment to the same script if these shapefiles turn out to be corrupted.

**Comment by julietcohen** [comment]
## Test run reveals parallelization errors with ray

Notes from most test run of the IWP workflow on 02/13:

**nodes:** 004 = head, 005 & 006 = workers

- had already removed (commented out) the `__start_tracking` and `__end_tracking` from `viz-raster` package because the ID object that is created by `__start_tracking` and is used by `__end_tracking` was erroring in past test runs (likely because the worker nodes were not working while the head node was, so the ID object was empty)
- configured the `slurm` script to output the successful ray output to a file (stdout.txt) that is separate from the ray output that indicates errors (stderr.txt), such as worker nodes failing. 

```
## output slurm info to a separate file than errors
#SBATCH -e stderr.txt
#SBATCH -o stdout.txt
## append output in info & error files rather than replacing last output
#SBATCH --open-mode=append
```

- stderr.txt does not include time stamps but I imagine the failures in the worker nodes 005 and 006 noted in this file occurred at the start of raster highest, since staging executed well on all 3 nodes and the rsync for raster highest worked with all 3 nodes.

<details>
  <summary>stdout.txt</summary>

```
This is BEST_cpu_ray_double_srun.slurm
IP Head: 172.28.22.67:6379
Starting HEAD at cn004
2023-02-13 17:25:22,341	INFO scripts.py:1060 -- Did not find any active Ray processes.
2023-02-13 17:25:24,881	INFO usage_lib.py:516 -- Usage stats collection is enabled by default without user confirmation because this terminal is detected to be non-interactive. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https://docs.ray.io/en/master/cluster/usage-stats.html for more details.
2023-02-13 17:25:24,881	INFO scripts.py:702 -- Local node IP: 172.28.22.67
2023-02-13 17:25:30,608	SUCC scripts.py:739 -- --------------------
2023-02-13 17:25:30,608	SUCC scripts.py:740 -- Ray runtime started.
2023-02-13 17:25:30,608	SUCC scripts.py:741 -- --------------------
2023-02-13 17:25:30,608	INFO scripts.py:743 -- Next steps
2023-02-13 17:25:30,609	INFO scripts.py:744 -- To connect to this Ray runtime from another node, run
2023-02-13 17:25:30,609	INFO scripts.py:747 --   ray start --address='172.28.22.67:6379'
2023-02-13 17:25:30,609	INFO scripts.py:763 -- Alternatively, use the following Python code:
2023-02-13 17:25:30,609	INFO scripts.py:765 -- import ray
2023-02-13 17:25:30,609	INFO scripts.py:769 -- ray.init(address='auto', _node_ip_address='172.28.22.67')
2023-02-13 17:25:30,609	INFO scripts.py:781 -- To connect to this Ray runtime from outside of the cluster, for example to
2023-02-13 17:25:30,609	INFO scripts.py:785 -- connect to a remote cluster from your laptop directly, use the following
2023-02-13 17:25:30,609	INFO scripts.py:789 -- Python code:
2023-02-13 17:25:30,609	INFO scripts.py:791 -- import ray
2023-02-13 17:25:30,609	INFO scripts.py:792 -- ray.init(address='ray://<head_node_ip_address>:10001')
2023-02-13 17:25:30,609	INFO scripts.py:801 -- To see the status of the cluster, use
2023-02-13 17:25:30,609	INFO scripts.py:802 --   ray status
2023-02-13 17:25:30,609	INFO scripts.py:812 -- If connection fails, check your firewall settings and network configuration.
2023-02-13 17:25:30,609	INFO scripts.py:820 -- To terminate the Ray runtime, run
2023-02-13 17:25:30,609	INFO scripts.py:821 --   ray stop
2023-02-13 17:25:30,609	INFO scripts.py:899 -- --block
2023-02-13 17:25:30,609	INFO scripts.py:900 -- This command will now block forever until terminated by a signal.
2023-02-13 17:25:30,609	INFO scripts.py:903 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
2023-02-13 17:25:53,763	INFO scripts.py:1060 -- Did not find any active Ray processes.
Starting WORKER 1 at cn005
WORKER 1 has SLURM_CPUS_PER_TASK: 
2023-02-13 17:25:56,835	INFO scripts.py:877 -- Local node IP: 172.28.22.68
2023-02-13 17:25:59,806	SUCC scripts.py:889 -- --------------------
2023-02-13 17:25:59,806	SUCC scripts.py:890 -- Ray runtime started.
2023-02-13 17:25:59,806	SUCC scripts.py:891 -- --------------------
2023-02-13 17:25:59,806	INFO scripts.py:893 -- To terminate the Ray runtime, run
2023-02-13 17:25:59,807	INFO scripts.py:894 --   ray stop
2023-02-13 17:25:59,807	INFO scripts.py:899 -- --block
2023-02-13 17:25:59,807	INFO scripts.py:900 -- This command will now block forever until terminated by a signal.
2023-02-13 17:25:59,807	INFO scripts.py:903 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
2023-02-13 17:26:24,523	INFO scripts.py:1060 -- Did not find any active Ray processes.
Starting WORKER 2 at cn006
WORKER 2 has SLURM_CPUS_PER_TASK: 
2023-02-13 17:26:27,649	INFO scripts.py:877 -- Local node IP: 172.28.22.69
2023-02-13 17:26:31,035	SUCC scripts.py:889 -- --------------------
2023-02-13 17:26:31,035	SUCC scripts.py:890 -- Ray runtime started.
2023-02-13 17:26:31,035	SUCC scripts.py:891 -- --------------------
2023-02-13 17:26:31,035	INFO scripts.py:893 -- To terminate the Ray runtime, run
2023-02-13 17:26:31,035	INFO scripts.py:894 --   ray stop
2023-02-13 17:26:31,035	INFO scripts.py:899 -- --block
2023-02-13 17:26:31,035	INFO scripts.py:900 -- This command will now block forever until terminated by a signal.
2023-02-13 17:26:31,035	INFO scripts.py:903 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
```
</details>

<details>
  <summary>stderr.txt</summary>

```
+ echo 'This is BEST_cpu_ray_double_srun.slurm'
+ export RAY_worker_register_timeout_seconds=120
+ RAY_worker_register_timeout_seconds=120
++ scontrol show hostnames 'cn[004-006]'
+ nodes='cn004
cn005
cn006'
+ nodes_array=($nodes)
+ head_node=cn004
++ srun --nodes=1 --ntasks=1 -w cn004 hostname --ip-address
+ head_node_ip=172.28.22.67
+ printf '%s\n' cn004 cn005 cn006
+ [[ 172.28.22.67 == *\ * ]]
+ port=6379
+ ip_head=172.28.22.67:6379
+ export ip_head
+ echo 'IP Head: 172.28.22.67:6379'
+ source /scratch/bbou/julietcohen/venv/iwp_2/bin/activate
++ deactivate nondestructive
++ '[' -n '' ']'
++ '[' -n '' ']'
++ '[' -n /bin/bash -o -n '' ']'
++ hash -r
++ '[' -n '' ']'
++ unset VIRTUAL_ENV
++ '[' '!' nondestructive = nondestructive ']'
++ VIRTUAL_ENV=/scratch/bbou/julietcohen/venv/iwp_2
++ export VIRTUAL_ENV
++ _OLD_VIRTUAL_PATH=/scratch/bbou/julietcohen/venv/iwp_2/bin:/u/julietcohen/.vscode-server/bin/441438abd1ac652551dbe4d408dfcec8a499b8bf/bin/remote-cli:/u/julietcohen/.cargo/bin:/u/julietcohen/.local/bin:/u/julietcohen/bin:/sw/spack/delta-2022-03/apps/cuda/11.6.1-gcc-11.2.0-vglutoe/bin:/sw/spack/delta-2022-03/apps/openmpi/4.1.2-gcc-11.2.0-37px7gc/bin:/sw/spack/delta-2022-03/apps/ucx/1.11.2-gcc-11.2.0-pymppfm/bin:/sw/spack/delta-2022-03/apps/gcc/11.2.0-gcc-8.4.1-fxgnsyr/bin:/sw/user/scripts:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/ddn/ime/bin:/opt/puppetlabs/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin
++ PATH=/scratch/bbou/julietcohen/venv/iwp_2/bin:/scratch/bbou/julietcohen/venv/iwp_2/bin:/u/julietcohen/.vscode-server/bin/441438abd1ac652551dbe4d408dfcec8a499b8bf/bin/remote-cli:/u/julietcohen/.cargo/bin:/u/julietcohen/.local/bin:/u/julietcohen/bin:/sw/spack/delta-2022-03/apps/cuda/11.6.1-gcc-11.2.0-vglutoe/bin:/sw/spack/delta-2022-03/apps/openmpi/4.1.2-gcc-11.2.0-37px7gc/bin:/sw/spack/delta-2022-03/apps/ucx/1.11.2-gcc-11.2.0-pymppfm/bin:/sw/spack/delta-2022-03/apps/gcc/11.2.0-gcc-8.4.1-fxgnsyr/bin:/sw/user/scripts:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/ddn/ime/bin:/opt/puppetlabs/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin:/opt/ddn/ime/bin
++ export PATH
++ '[' -n '' ']'
++ '[' -z '' ']'
++ _OLD_VIRTUAL_PS1=
++ PS1='(iwp_2) '
++ export PS1
++ '[' -n /bin/bash -o -n '' ']'
++ hash -r
+ export RAY_max_pending_lease_requests_per_scheduling_category=2000
+ RAY_max_pending_lease_requests_per_scheduling_category=2000
+ echo 'Starting HEAD at cn004'
+ srun --nodes=1 --ntasks=1 -w cn004 ray stop
+ sleep 1
+ srun --nodes=1 --ntasks=1 -w cn004 ray start --head --node-ip-address=172.28.22.67 --port=6379 --dashboard-host 0.0.0.0 --log-color true --block
+ worker_num=2
+ (( i = 1 ))
+ (( i <= worker_num ))
+ node_i=cn005
+ srun --nodes=1 --ntasks=1 -w cn005 ray stop
+ echo 'Starting WORKER 1 at cn005'
+ srun --nodes=1 --ntasks=1 -w cn005 ulimit -n unlimited
/usr/bin/ulimit: line 2: ulimit: open files: cannot modify limit: Operation not permitted
srun: error: cn005: task 0: Exited with exit code 1
+ srun --nodes=1 --ntasks=1 -w cn005 export RAY_worker_register_timeout_seconds=120
slurmstepd: error: execve(): export: No such file or directory
srun: error: cn005: task 0: Exited with exit code 2
+ echo 'WORKER 1 has SLURM_CPUS_PER_TASK: '
+ (( i++ ))
+ (( i <= worker_num ))
+ srun --nodes=1 --ntasks=1 -w cn005 ray start --address 172.28.22.67:6379 --log-color true --block
+ node_i=cn006
+ srun --nodes=1 --ntasks=1 -w cn006 ray stop
[2023-02-13 17:25:59,804 I 453213 453213] global_state_accessor.cc:357: This node has an IP address of 172.28.22.68, while we can not find the matched Raylet address. This maybe come from when you connect the Ray cluster with a different IP address or connect a container.
+ echo 'Starting WORKER 2 at cn006'
+ srun --nodes=1 --ntasks=1 -w cn006 ulimit -n unlimited
/usr/bin/ulimit: line 2: ulimit: open files: cannot modify limit: Operation not permitted
srun: error: cn006: task 0: Exited with exit code 1
+ srun --nodes=1 --ntasks=1 -w cn006 export RAY_worker_register_timeout_seconds=120
slurmstepd: error: execve(): export: No such file or directory
srun: error: cn006: task 0: Exited with exit code 2
+ echo 'WORKER 2 has SLURM_CPUS_PER_TASK: '
+ srun --nodes=1 --ntasks=1 -w cn006 ray start --address 172.28.22.67:6379 --log-color true --block
+ (( i++ ))
+ (( i <= worker_num ))
+ sleep infinity
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
slurmstepd: error: *** STEP 1396073.2 ON cn004 CANCELLED AT 2023-02-13T18:25:03 DUE TO TIME LIMIT ***
slurmstepd: error: *** STEP 1396073.10 ON cn006 CANCELLED AT 2023-02-13T18:25:03 DUE TO TIME LIMIT ***
slurmstepd: error: *** JOB 1396073 ON cn004 CANCELLED AT 2023-02-13T18:25:03 DUE TO TIME LIMIT ***
slurmstepd: error: *** STEP 1396073.6 ON cn005 CANCELLED AT 2023-02-13T18:25:03 DUE TO TIME LIMIT ***
```
</details>

 
- transferring footprints to nodes went well on all nodes
- staging worked well in parallel, wrote files to all 3 nodes `/tmp`, transferred to `/scratch`, and merged into 214 staged files
- raster highest step went well, **but only the head node processed all the 214 files into geotiffs**, wrote to `/tmp` on node 004 then transferred files to `/scratch` succesfully
- raster lower went well, but had to configure this step to read and write all files from/to `/scratch` so all lower z-levels could read all files in the z-level above (z-level 14 cannot be dispersed across 3 nodes' `/tmp` directories because then z-level 13 does not have access to all those files). The 214 highest z-level geotiffs became 1076 geotiffs in total, stored in `/scratch/bbou/julietcohen/IWP/output/iwp_testRun_20230213/geotiff`
- **did not write the summary csv files** `raster_summary.csv` and `rasterization_events.csv` (even tho the paths to write these were included in the config) because these files are written by `__end_tracking` (see [here](https://github.com/PermafrostDiscoveryGateway/viz-raster/blob/a12c9b732c3e7d2280ba68ade7de94ffc5512f2e/pdgraster/RasterTiler.py#L664)) which was not part of the rasterization process in this run
- as a result of not having a `rasters_summary.csv`, the web tiling could not update ranges. I adjusted `IN_PROGRESS_VIZ_WORKFLOW.py` so it wouldn't try that part, and just wanted to see if web tiling would execute anyway, but ran into a different error: the step was not able to execute `geotiff_paths = rasterizer.tiles.get_filenames_from_dir(base_dir = 'geotiff_remote')` because the base dir was not successfully created above. This error is separate from the worker nodes not being functional in the rasterization steps.
- The version of  `IN_PROGRESS_VIZ_WORKFLOW.py` that was used for this run is [here](https://github.com/PermafrostDiscoveryGateway/viz-workflow/blob/414d98e8eb63db25b34821f21804471af95414b5/IN_PROGRESS_VIZ_WORKFLOW.py).

**Comment by julietcohen** [comment]
## Test Run 02/15 staging

- removed [these lines](https://github.com/PermafrostDiscoveryGateway/viz-workflow/blob/eeb58466849c7eb7f1ed927113d44c7c7fdcfc02/slurm/BEST_cpu_ray_double_srun.slurm#L96) from the slurm script that sets the node limit for number of open files to unlimited:
```
    srun --nodes=1 --ntasks=1 -w "$node_i" \
      ulimit -n unlimited`
```
because of the error that that was not allowed was printed in stderr.txt in the last run.
- changed slurm from `--exclusive` to `--mem=200g` to try to unblock the process from starting, was receiveing blocks of (Resources) and (Priority)
- this run was **much** slower than previous runs, perhaps due to the change from `--exclusive` to `--mem=200g`, or because Delta just has a lot of users running jobs now?
- Staging used to take just a couple minutes in past runs with 3 CPU nodes, this time took the almost the full 2 hours. 
- Occasionally an obscure message was printed in the terminal, examples:
![image](https://user-images.githubusercontent.com/88209596/219222286-3c771cd0-c823-465b-9d1f-efa971acbca5.png)
![message](https://user-images.githubusercontent.com/88209596/219222568-6ea7b737-dd89-4b12-8658-295ee474e7be.png)

- Froze at file 250/279 for a long time, then output this:
![raylet](https://user-images.githubusercontent.com/88209596/219233845-028e583f-7954-4c36-b5dd-e07294e5ea8d.png)

- While staging was running, checked the stdout.txt and stderr.txt. The stdout.txt looks fine. the stderr.txt does not contain the `ulimit -n unlimited` error reported in the last run, but still contains the error:

```
+ node_i=cn041
+ srun --nodes=1 --ntasks=1 -w cn041 ray stop
+ srun --nodes=1 --ntasks=1 -w cn039 ray start --address 172.28.22.65:6379 --log-color true --block
[2023-02-15 17:33:08,680 I 1227975 1227975] global_state_accessor.cc:357: This node has an IP address of 172.28.22.102, while we can not find the matched Raylet address. This maybe come from when you connect the Ray cluster with a different IP address or connect a container.
+ echo 'Starting WORKER 2 at cn041'
+ srun --nodes=1 --ntasks=1 -w cn041 export RAY_worker_register_timeout_seconds=120
srun: error: cn041: task 0: Exited with exit code 2
slurmstepd: error: execve(): export: No such file or directory
+ echo 'WORKER 2 has SLURM_CPUS_PER_TASK: '
```

<details>
  <summary>stdout.txt</summary>

```

```
</details>

<details>
  <summary>stderr.txt</summary>

```

```
</details>

However, this did not interfere with parallelization during staging; `glances` in each node showed that CPU usage was around 20% for each, and it would be <1% if it was not processing. Additionally, the staged files were written to all 3 nodes' `/tmp` directories (just like past runs), so the errors that are reported in stderr.txt may not be the root of what is causing the errors in raster parallel processing?

Head node:
![image](https://user-images.githubusercontent.com/88209596/219231223-67e0c104-f298-44b7-8ddf-d063c1597188.png)

Worker node:
![image](https://user-images.githubusercontent.com/88209596/219231286-23859115-24e2-4437-936c-82633f424142.png)

Other worker node:
![image](https://user-images.githubusercontent.com/88209596/219231348-44e8d606-8203-4447-a98d-3025b27b3054.png)

Didn't reach rasterization because staging took so long.

**Comment by julietcohen** [comment]
Note on the last ray workflow run:
- Kastan pointed out that the CPU usage could refer to computation for other jobs (such as for user qin3 listed in the screenshots). So it is unclear how much CPU was being allocated to my job specifically. The only indication that processing was ocurring on all 3 nodes was the files being written to all 3 nodes' `/scratch` dirs.
- Moving forward, I will use the ray dashboard: -L localhost:8265:cn005.delta.internal.ncsa.edu:8265 [dt-login.delta.ncsa.illinois.edu](http://dt-login.delta.ncsa.illinois.edu/) (replacing cn005 with my node) on my local laptop to monitor node activity
- The line in the slurm file `export RAY_worker_register_timeout_seconds=120` is not trying to write a file, it is just allowing the nodes more time to initiate than the default, so still not sure why that error is occurring.
- I am interested in scheduling a job on Delta in advance to try to snag nodes with higher efficiency

**Comment by julietcohen** [comment]
## First Test Run 02/17 (Staging only)

Removing the timeout line in the slurm script resulted in this output during staging, which matches Kastan's description of his experience before adding in the timeout to the slurm script:

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/88209596/219767405-a369b320-8607-4a47-a71c-aa791836c1eb.png">

The raylet connection error is still present in stderr.txt fro both worker nodes:

```
[2023-02-17 13:10:10,835 I 525378 525378] global_state_accessor.cc:357: This node has an IP address of 172.28.22.160, while we can not find the matched Raylet address. This maybe come from when you connect the Ray cluster with a different IP address or connect a container.
```

Staged files are written way too slowly again.
- Strange output during staging I haven't seen before:

<img width="452" alt="image" src="https://user-images.githubusercontent.com/88209596/219768320-6d52ce6b-bd7b-4de4-bf1c-14e625aa2c2e.png">

- glances reveals that my CPU is 0.2% (only my processes are listed this time, so the 0.2% usage definitely refers to my job)
<img width="1295" alt="image" src="https://user-images.githubusercontent.com/88209596/219780975-76ec1420-7098-48e1-8307-6d8128197889.png">

Staging was becoming slower and slower. Cancelled the job.

## Second Test Run 02/17 (all 4 steps through web tiling)

Success! Used a workaround in `RasterTiler.py` and one more small manual step (manually creating `geotiff` and `web_tile` dirs in `/scratch`) to get the raster step to write the `raster_summary.csv` file so the ranges can be updated in web tiling so the colors are correct per z-level. Need to check the colors still using Cesium map or upload to demo portal. 

CPU usage was up to 40%!

Also tested staging files for all of Alaska dir in high ice and it only took a few minutes. Rasterization timeline for that many tiles is TBD.
